storage/aptosdb/src/pruner/state_store/mod.rs

```diff
@@ -102,140 +86,112 @@ impl DBPruner for StateStorePruner {
-        if now - *self.index_purged_at.lock() > MIN_INTERVAL
-            && min_readable_version - max_index_purged_version > MIN_VERSIONS
-        {
-            self.db.range_delete::<StaleNodeIndexSchema, Version>(
-                &(max_index_purged_version + 1), // begin is inclusive
-                &(min_readable_version + 1),     // end is exclusive
-            )?;
-            // The index items at min_readable_version has been consumed already because they
-            // indicate nodes that became stale at that version, keeping that version readable,
-            // hence is purged above.
-            self.max_index_purged_version
-                .store(min_readable_version, Ordering::Relaxed);
-            *self.index_purged_at.lock() = now;
-        }
-        Ok(())
-    }
-
-    pub fn target_version(&self) -> Version {
-        self.target_version.load(Ordering::Relaxed)
-    }
-}
-
-pub fn prune_state_store(
-    db: &DB,
-    min_readable_version: Version,
-    target_version: Version,
-    max_versions: usize,
-) -> Result<Version> {
-    let indices =
-        StaleNodeIndicesByVersionIterator::new(db, min_readable_version + 1, target_version)?
-            .take(max_versions) // Iterator<Item = Result<Vec<StaleNodeIndex>>>
-            .collect::<Result<Vec<_>>>()? // now Vec<Vec<StaleNodeIndex>>
-            .into_iter()
-            .flatten()
-            .collect::<Vec<_>>();
-
-    if indices.is_empty() {
-        Ok(target_version)
-    } else {
-        let _timer = OTHER_TIMERS_SECONDS
-            .with_label_values(&["pruner_commit"])
-            .start_timer();
-        let new_min_readable_version = indices.last().expect("Should exist.").stale_since_version;
-        let mut batch = SchemaBatch::new();
-        indices
-            .into_iter()
-            .try_for_each(|index| batch.delete::<JellyfishMerkleNodeSchema>(&index.node_key))?;
-        db.write_schemas(batch)?;
-        Ok(new_min_readable_version)
-    }
-}
-
-struct StaleNodeIndicesByVersionIterator<'a> {
-    inner: Peekable<SchemaIterator<'a, StaleNodeIndexSchema>>,
-    target_min_readable_version: Version,
-}
-
-impl<'a> StaleNodeIndicesByVersionIterator<'a> {
-    fn new(
-        db: &'a DB,
+    pub fn prune_state_store(
+        &self,
         min_readable_version: Version,
-        target_min_readable_version: Version,
-    ) -> Result<Self> {
-        let mut iter = db.iter::<StaleNodeIndexSchema>(ReadOptions::default())?;
-        iter.seek(&min_readable_version)?;
-
-        Ok(Self {
-            inner: iter.peekable(),
-            target_min_readable_version,
-        })
+        target_version: Version,
+        batch_size: usize,
+    ) -> anyhow::Result<Version> {
+        assert_ne!(batch_size, 0);
+        let (indices, is_end_of_target_version) =
+            self.get_stale_node_indices(min_readable_version, target_version, batch_size)?;
+        if indices.is_empty() {
+            self.is_pruning_pending
```

Comment:
hmm. will it work if we record "completed_processed_version" instead of "min_readable_version" which is "version that we started processing"?

Reply:
Seems that we still need `min_readble_version` for this function https://github.com/aptos-labs/aptos-core/blob/581fb4754e7f21288733bce29aa7d2817ba39688/storage/aptosdb/src/lib.rs#L168. Are you suggesting we can deduce the `min_readble_version` from `completed_processed_version`? What is your concern with the current implementation?

Reply:
Added  completed_processed_version instead.